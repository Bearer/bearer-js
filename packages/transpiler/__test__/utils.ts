import * as ts from 'typescript'
export default (sourceFile: ts.SourceFile) => {
  let names: string[] = []
  traverseNode(sourceFile, names)
  function traverseNode(node: ts.Node, names: string[]) {
    if (node.kind === ts.SyntaxKind.Decorator) names.push(node.expression.expression.escapedText)
    ts.forEachChild(node, node => {
      traverseNode(node, names)
    })
  }
  return names
}
