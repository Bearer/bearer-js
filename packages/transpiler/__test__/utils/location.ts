import * as path from 'path'

export const FIXTURES = path.join(__dirname, '..', '__fixtures__')
export const BUILD = path.join(__dirname, '..', '.build/src')

export function UnitFixtureDirectory(dir: string) {
  return path.join(FIXTURES, 'unit', dir)
}

export function BuildUnitFixtureDirectory(dir: string) {
  return path.join(BUILD, 'unit', dir)
}
