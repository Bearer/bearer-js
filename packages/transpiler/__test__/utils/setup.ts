import * as fs from 'fs'
import * as globby from 'globby'
import * as path from 'path'

const buildFolder = path.join(__dirname, '..', '.build/src')

process.env.BEARER_SCENARIO_ID = 'SPONGE_BOB'

globby.sync(['**/*.tsx', '**/*.ts', '**/*.json'], { cwd: buildFolder }).forEach(file => {
  const filePath = path.join(buildFolder, file)
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath)
  }
})
