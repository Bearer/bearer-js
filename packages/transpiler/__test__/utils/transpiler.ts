import Transpiler, { TranpilerOptions } from '../../src/index'

export function TranspilerFactory(options: Partial<TranpilerOptions>): Transpiler {
  const defaults: TranpilerOptions = {
    watchFiles: false,
    buildFolder: '../../../.build/'
  }

  return new Transpiler({
    ...defaults,
    ...options
  })
}
