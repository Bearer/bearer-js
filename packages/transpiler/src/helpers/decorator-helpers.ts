import * as ts from 'typescript'

export function hasPropDecoratedWithName(classNode: ts.ClassDeclaration, decoratorName: string): boolean {
  return Boolean(propDecoratedWithName(classNode, decoratorName).length)
}

export function propDecoratedWithName(
  node: ts.ClassDeclaration,
  decoratorName: string
): Array<ts.PropertyDeclaration | null> {
  const props: Array<ts.PropertyDeclaration | null> = []
  ts.forEachChild(node, node => {
    if (ts.isPropertyDeclaration(node) && hasDecoratorNamed(node as ts.PropertyDeclaration, decoratorName)) {
      props.push(node)
    }
  })
  return props
}

export function hasDecoratorNamed(
  decoratedNode: ts.PropertyDeclaration | ts.ClassDeclaration | ts.MethodDeclaration,
  name: string
): boolean {
  return ts.forEachChild(decoratedNode, anode => {
    return ts.isDecorator(anode) && decoratorNamed(anode, name)
  })
}

export function decoratorNamed(tsDecorator: ts.Decorator, name: string): boolean {
  return ts.forEachChild(tsDecorator, node => {
    return ts.isCallExpression(node) && (node.expression as ts.Identifier).escapedText === name
  })
}

export function getDecoratorProperties(tsDecorator: ts.Decorator, index = 0): ts.ObjectLiteralExpression {
  if (!ts.isCallExpression(tsDecorator.expression)) {
    return ts.createObjectLiteral([])
  }
  const expression = tsDecorator.expression
  const argument = expression.arguments[index]

  if (!ts.isObjectLiteralExpression(argument)) {
    return ts.createObjectLiteral([])
  }
  return ts.createObjectLiteral(argument.properties)
}

export function getExpressionFromLiteralObject<T extends ts.Expression>(
  tsLiteral: ts.ObjectLiteralExpression,
  key: string
): T {
  const found: ts.PropertyAssignment = tsLiteral.properties.find(property => {
    if (ts.isStringLiteral(property.name)) {
      return property.name.text === key
    } else {
      return (property.name as ts.Identifier).escapedText.toString() === key
    }
  }) as ts.PropertyAssignment
  if (found) {
    return found.initializer as T
  }
  return null
}

export function getExpressionFromDecorator<T extends ts.Expression>(
  tsDecorator: ts.Decorator,
  key: string,
  index = 0
): T {
  return getExpressionFromLiteralObject(getDecoratorProperties(tsDecorator, index), key)
}

export function getDecoratorNamed(
  tsDecoratedNode: { decorators?: ts.NodeArray<ts.Decorator> },
  name: string
): ts.Decorator | undefined {
  if (tsDecoratedNode.decorators) {
    return tsDecoratedNode.decorators.find(decorator => decoratorNamed(decorator, name))
  }
}

// tslint:disable-next-line:prefer-array-literal
export function extractStringOptions<T>(objectLitteral: ts.ObjectLiteralExpression, keys: Array<keyof T>): Partial<T> {
  const values: Partial<T> = {}
  keys.map(value => {
    const optionValue = getExpressionFromLiteralObject<ts.StringLiteral>(objectLitteral, value as string)
    if (optionValue) {
      values[value as string] = optionValue && optionValue.text
    }
  })
  return values
}

export function extractArrayOptions<T>(
  objectLitteral: ts.ObjectLiteralExpression,
  keys: (keyof T)[]
): { [P in keyof T]?: string[] } {
  const values = {}
  keys.map(value => {
    const optionValue = getExpressionFromLiteralObject<ts.StringLiteral>(objectLitteral, value as string)
    if (optionValue && ts.isArrayLiteralExpression(optionValue)) {
      values[value as string] = optionValue.elements.map(e => {
        if (ts.isStringLiteral(e)) {
          return e.text
        }
        return []
      })
    }
  })
  return values
}

export function extractBooleanOptions<T>(objectLitteral: ts.ObjectLiteralExpression, keys: (keyof T)[]): Partial<T> {
  const values: Partial<T> = {}
  keys.map(value => {
    const optionValue = getExpressionFromLiteralObject<ts.BooleanLiteral>(objectLitteral, value as string)
    if (optionValue) {
      values[value as string] = optionValue && optionValue.kind === ts.SyntaxKind.TrueKeyword
    }
  })

  return values
}
