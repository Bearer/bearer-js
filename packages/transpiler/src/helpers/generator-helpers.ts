import * as ts from 'typescript'

import { createOrUpdateComponentDidLoad } from '../../src/transformers/bearer'

import { Decorators } from './../constants'
import { CreateFetcherMeta, TAddAutoLoad, TCreateLoadDataCall, TCreateLoadResourceMethod } from './../types'
import { idName } from './name-helpers'

export function createFetcher(meta: CreateFetcherMeta) {
  return ts.createProperty(
    [
      ts.createDecorator(
        ts.createCall(ts.createIdentifier(Decorators.Intent), undefined, [ts.createLiteral(meta.intentName)])
      )
    ],
    undefined,
    meta.intentMethodName,
    undefined,
    undefined,
    undefined
  )
}

function propertyReferenceIdNames(meta: TCreateLoadResourceMethod, metaCollection: TCreateLoadResourceMethod[]) {
  return meta.intentArguments.map(name => {
    const metaInfo = metaCollection.find(meta => meta.propDeclarationName === name)
    if (metaInfo) {
      return ts.createPropertyAssignment(
        idName(name),
        ts.createPropertyAccess(ts.createThis(), metaInfo.propertyReferenceIdName)
      )
    }
    return ts.createPropertyAssignment(name, ts.createPropertyAccess(ts.createThis(), name))
  })
}

export function createLoadResourceMethod(meta: TCreateLoadResourceMethod, metaCollection: TCreateLoadResourceMethod[]) {
  const intentCall = ts.createCall(ts.createPropertyAccess(ts.createThis(), meta.intentMethodName), undefined, [
    ts.createObjectLiteral([
      ts.createPropertyAssignment(
        meta.intentReferenceIdKeyName,
        ts.createPropertyAccess(ts.createThis(), meta.propertyReferenceIdName)
      ),
      ...propertyReferenceIdNames(meta, metaCollection)
    ])
  ])
  const udapteState = ts.createArrowFunction(
    undefined,
    undefined,
    [
      ts.createParameter(
        undefined,
        undefined,
        undefined,
        ts.createObjectBindingPattern([ts.createBindingElement(undefined, undefined, 'data')]),
        undefined,
        ts.createTypeLiteralNode([
          ts.createPropertySignature(undefined, 'data', undefined, meta.typeIdentifier, undefined)
        ]),
        undefined
      )
    ],
    undefined,
    undefined,
    ts.createBlock(
      [
        ts.createIf(
          ts.createIdentifier('data'),
          ts.createBlock([
            ts.createStatement(
              ts.createBinary(
                ts.createPropertyAccess(ts.createThis(), meta.propDeclarationName),
                ts.SyntaxKind.EqualsToken,
                ts.createIdentifier('data')
              )
            )
          ])
        )
      ],
      true
    )
  )
  const promiseHandler = ts.createCall(ts.createPropertyAccess(intentCall, 'then'), undefined, [udapteState])
  return ts.createProperty(
    undefined,
    undefined,
    meta.loadMethodName,
    undefined,
    undefined,
    ts.createArrowFunction(
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      ts.createBlock([ts.createStatement(promiseHandler)], true)
    )
  )
}

export function createLoadDataCall(meta: TCreateLoadDataCall) {
  return ts.createStatement(
    ts.createCall(ts.createPropertyAccess(ts.createThis(), meta.loadMethodName), undefined, undefined)
  )
}

export function addAutoLoad(tsClass: ts.ClassDeclaration, meta: TAddAutoLoad): ts.ClassDeclaration {
  if (meta.autoLoad) {
    return createOrUpdateComponentDidLoad(tsClass, block =>
      ts.updateBlock(block, [
        ...block.statements,
        ts.createIf(
          ts.createPropertyAccess(ts.createThis(), meta.propertyReferenceIdName),
          ts.createBlock([createLoadDataCall(meta)])
        )
      ])
    )
  }
  return tsClass
}
