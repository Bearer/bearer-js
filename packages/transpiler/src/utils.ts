import * as ts from 'typescript'

export function getSourceCode(node): string {
  const resultFile = ts.createSourceFile(
    'tmp.ts',
    '',
    ts.ScriptTarget.Latest,
    /*setParentNodes*/ false,
    ts.ScriptKind.TS
  )
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed
  })

  return printer.printNode(ts.EmitHint.Unspecified, node, resultFile)
}
