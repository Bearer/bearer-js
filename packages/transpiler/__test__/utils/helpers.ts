import fs from 'fs'
import path from 'path'

import { TranpilerOptions } from '../../src/index'
import { BuildUnitFixtureDirectory, UnitFixtureDirectory } from '../utils/location'

import { TranspilerFactory } from './transpiler'

const fixtures = path.join(__dirname, '__fixtures__')

export function runUnitOn(name: string, transpilerOptions: Partial<TranpilerOptions> = {}) {
  const srcDirectory = UnitFixtureDirectory(name)
  const buildDirectory = BuildUnitFixtureDirectory(name)

  beforeAll(() => {
    runTranspiler(transpilerOptions)
  })

  fs.readdirSync(srcDirectory).forEach(file => {
    describe(file.replace(/\.tsx?/, '').replace(/\-/g, ' '), () => {
      it('match snapshot', async done => {
        expect.assertions(1)
        fs.readFile(path.join(buildDirectory, file), 'utf8', (_e, postContent) => {
          expect({
            postContent,
            file
          }).toMatchSnapshot()
          done()
        })
      })
    })
  })
}

export function runTranspiler(transpilerOptions: Partial<TranpilerOptions> = {}) {
  const transpiler = TranspilerFactory({
    ...transpilerOptions,
    ROOT_DIRECTORY: fixtures,
    srcFolder: '../../__fixtures__'
  })

  transpiler.run()
}
