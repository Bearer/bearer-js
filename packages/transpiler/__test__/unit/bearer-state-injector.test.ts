import { runUnitOn } from '../utils/helpers'
const TEST_NAME = 'bearer-state-injector'

describe('Bearer State Injector', () => {
  runUnitOn(TEST_NAME)
})
