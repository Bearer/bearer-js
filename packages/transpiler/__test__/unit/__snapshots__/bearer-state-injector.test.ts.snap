// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bearer State Injector do not process match snapshot 1`] = `
Object {
  "file": "do-not-process.ts",
  "postContent": "class DoNotProcess {
    constructor() { }
}
",
}
`;

exports[`Bearer State Injector do not rewatch property match snapshot 1`] = `
Object {
  "file": "do-not-rewatch-property.tsx",
  "postContent": "import { BearerState, Watch, State } from '@bearer/core';
class UpdateExistingPropertyWatcher {
    @BearerState()
    @State()
    pullRequests: Array<any> = [];
    @BearerState({ statePropName: 'repo' })
    @State()
    repository: {} = {};
    constructor() { }
    @Prop({ \\"context\\": \\"bearer\\" })
    bearerContext: any;
    @Watch('pullRequests')
    pullRequestsChangeHandler(newPullRequest: Array<any>, oldValue: any[]) { this.bearerContext.update(\\"pullRequests\\", newPullRequest); console.log('Prepend stuff'); }
    @Watch(\\"repository\\")
    _notifyBearerStateHandler_repository(newValue: any) { this.bearerContext.update(\\"repo\\", newValue); }
    componentWillLoad() { this.bearerContext.subscribe(this); }
    componentDidUnload() { this.bearerContext.unsubscribe(this); }
    bearerUpdateFromState = state => { this.pullRequests = state[\\"pullRequests\\"]; this.repository = state[\\"repo\\"]; };
}
",
}
`;

exports[`Bearer State Injector does not re add component lifecycle methods match snapshot 1`] = `
Object {
  "file": "does-not-re-add-component-lifecycle-methods.tsx",
  "postContent": "import { BearerState, Watch, State } from '@bearer/core';
class UpdateExistingComponentLifecycleMethods {
    @BearerState()
    @State()
    pullRequests: Array<any> = [];
    // TODO: handle these use cases
    componentWillLoad() { this.bearerContext.subscribe(this); console.log('componentWillLoad'); }
    componentDidUnload() { this.bearerContext.unsubscribe(this); console.log('componentDidUnload'); }
    constructor() { }
    @Prop({ \\"context\\": \\"bearer\\" })
    bearerContext: any;
    @Watch(\\"pullRequests\\")
    _notifyBearerStateHandler_pullRequests(newValue: any) { this.bearerContext.update(\\"pullRequests\\", newValue); }
    bearerUpdateFromState = state => { this.pullRequests = state[\\"pullRequests\\"]; };
}
",
}
`;

exports[`Bearer State Injector ensure decorator injected match snapshot 1`] = `
Object {
  "file": "ensure-decorator-injected.ts",
  "postContent": "import { BearerState, Watch, State } from '@bearer/core';
class AddDecoratorOnTranspile {
    @BearerState()
    @State()
    pullRequests: Array<any> = [];
    constructor() { }
    @Prop({ \\"context\\": \\"bearer\\" })
    bearerContext: any;
    @Watch(\\"pullRequests\\")
    _notifyBearerStateHandler_pullRequests(newValue: any) { this.bearerContext.update(\\"pullRequests\\", newValue); }
    componentWillLoad() { this.bearerContext.subscribe(this); }
    componentDidUnload() { this.bearerContext.unsubscribe(this); }
    bearerUpdateFromState = state => { this.pullRequests = state[\\"pullRequests\\"]; };
}
",
}
`;

exports[`Bearer State Injector has bearer context present match snapshot 1`] = `
Object {
  "file": "has-bearer-context-present.tsx",
  "postContent": "import { BearerState, Prop, Watch, State } from '@bearer/core';
class HasBearerContextInjected {
    @BearerState()
    @State()
    attachPullRequests: any[] = [];
    @Prop({ context: 'bearer' })
    bearerContext: any;
    constructor() { }
    @Watch(\\"attachPullRequests\\")
    _notifyBearerStateHandler_attachPullRequests(newValue: any) { this.bearerContext.update(\\"attachPullRequests\\", newValue); }
    componentWillLoad() { this.bearerContext.subscribe(this); }
    componentDidUnload() { this.bearerContext.unsubscribe(this); }
    bearerUpdateFromState = state => { this.attachPullRequests = state[\\"attachPullRequests\\"]; };
}
",
}
`;
